File: src/splash/command.rs
use super::protocol::{Frame, Opcode, ProtocolError};
use super::root_transition::{DeferredRootTransition, RootTransition, RootTransitionError};
use super::state::{Lifecycle, Mode, RootStage, SplashState, StateAction, View};
use std::path::Path;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CommandOutcome {
    pub response: Frame,
    pub should_quit: bool,
    /// Keep the last splash pixels when the backend can do so without keeping
    /// terminal, keyboard, KD, or VT ownership.
    pub retain_splash: bool,
    /// A root transition failure. The daemon must restore presentation state
    /// and terminate rather than risk retaining the VT after its handoff unit
    /// has already allowed boot to continue. `rollback_incomplete()` further
    /// indicates that the process filesystem namespace is unknown.
    pub fatal_root_transition: Option<RootTransitionError>,
}

pub fn is_mutating(opcode: Opcode) -> bool {
    !matches!(opcode, Opcode::Ping | Opcode::State | Opcode::NativeReady)
}

pub fn handle_request(
    state: &mut SplashState,
    request: &Frame,
) -> Result<CommandOutcome, ProtocolError> {
    let mut transition = DeferredRootTransition;
    handle_request_with_root_transition(state, request, &mut transition)
}

pub fn handle_request_with_root_transition(
    state: &mut SplashState,
    request: &Frame,
    root_transition: &mut dyn RootTransition,
) -> Result<CommandOutcome, ProtocolError> {
    let request_id = request.request_id();
    let mut should_quit = false;
    let mut retain_splash = false;

    let action = match request.opcode() {
        Opcode::Ping => {
            return Ok(CommandOutcome {
                response: Frame::pong(request_id),
                should_quit: false,
                retain_splash: false,
                fatal_root_transition: None,
            });
        }
        Opcode::State => {
            return Ok(CommandOutcome {
                response: Frame::state_result(request_id, state_json(state))?,
                should_quit: false,
                retain_splash: false,
                fatal_root_transition: None,
            });
        }
        Opcode::NativeReady => {
            // Readiness is a daemon-scoped capability: only the event loop can
            // prove that the prepared native broker still has an enabled
            // coordinator. Pure command handling must never guess from socket
            // or filesystem presence.
            return Ok(CommandOutcome {
                response: Frame::error(
                    request_id,
                    "native password broker is unavailable in this command context",
                )?,
                should_quit: false,
                retain_splash: false,
                fatal_root_transition: None,
            });
        }
        Opcode::Show => Some(StateAction::Show),
        Opcode::Hide => Some(StateAction::Hide),
        Opcode::Status => {
            let text = request.payload_text()?;
            Some(StateAction::SetStatus(
                (!text.is_empty()).then(|| text.to_owned()),
            ))
        }
        Opcode::Progress => Some(StateAction::SetProgress(request.progress_value())),
        Opcode::Message => Some(StateAction::SetMessage(Some(
            request.payload_text()?.to_owned(),
        ))),
        Opcode::HideMessage => {
            let requested = request.payload_text()?;
            if requested.is_empty() || state.message() == Some(requested) {
                Some(StateAction::SetMessage(None))
            } else {
                None
            }
        }
        Opcode::DetailsShow => Some(StateAction::ShowDetails),
        Opcode::DetailsHide => Some(StateAction::HideDetails),
        Opcode::DetailsToggle => Some(StateAction::ToggleDetails),
        Opcode::Deactivate => Some(StateAction::Deactivate),
        Opcode::Reactivate => Some(StateAction::Reactivate),
        Opcode::SetMode => Some(StateAction::SetMode(
            request
                .mode_value()
                .expect("validated mode requests contain a mode"),
        )),
        Opcode::UpdateRootFs => {
            let validated_path = request.payload_text()?;
            if state.root_stage() == RootStage::Initramfs
                && let Err(error) = state.apply(StateAction::SetRootStage(RootStage::Switching))
            {
                let response = Frame::error(request_id, error.to_string())?;
                // The authenticated initramfs handoff job is best-effort and
                // boot is allowed to continue after it returns. Even a
                // pre-transition rejection (most importantly an active
                // prompt) must therefore stop presentation ownership: a later
                // real-root client is not guaranteed to remain able to reach
                // this initramfs namespace. The error response is deferred by
                // the daemon until display/runtime cleanup completes.
                let _ = state.apply(StateAction::FailOpen);
                return Ok(CommandOutcome {
                    response,
                    should_quit: true,
                    retain_splash: false,
                    fatal_root_transition: None,
                });
            }
            if state.root_stage() == RootStage::Switching {
                if let Err(error) = root_transition.transition(Path::new(validated_path)) {
                    let response = Frame::error(request_id, error.to_string())?;
                    // A fully rolled-back chroot is safe at the filesystem
                    // layer, but it is not safe to keep owning the display:
                    // init will continue past this best-effort handoff and no
                    // later real-root client is guaranteed to reach us. Every
                    // transition failure therefore takes the same fail-open
                    // restoration path. Incomplete rollback remains encoded in
                    // the error for diagnostics.
                    let _ = state.apply(StateAction::FailOpen);
                    return Ok(CommandOutcome {
                        response,
                        should_quit: true,
                        retain_splash: false,
                        fatal_root_transition: Some(error),
                    });
                }
                Some(StateAction::SetRootStage(RootStage::RealRoot))
            } else {
                None
            }
        }
        Opcode::Quit => {
            should_quit = true;
            retain_splash = request.retains_splash();
            Some(StateAction::Quit)
        }
        Opcode::Ack | Opcode::Error | Opcode::Pong | Opcode::StateResult => {
            return Ok(CommandOutcome {
                response: Frame::error(request_id, "response opcode is invalid in a request")?,
                should_quit: false,
                retain_splash: false,
                fatal_root_transition: None,
            });
        }
    };

    if let Some(action) = action
        && let Err(error) = state.apply(action)
    {
        return state_error(request_id, error);
    }

    Ok(CommandOutcome {
        response: Frame::ack(request_id),
        should_quit,
        retain_splash,
        fatal_root_transition: None,
    })
}

fn state_error(
    request_id: u64,
    error: impl std::fmt::Display,
) -> Result<CommandOutcome, ProtocolError> {
    Ok(CommandOutcome {
        response: Frame::error(request_id, error.to_string())?,
        should_quit: false,
        retain_splash: false,
        fatal_root_transition: None,
    })
}

pub fn state_json(state: &SplashState) -> String {
    let lifecycle = match state.lifecycle() {
        Lifecycle::Starting => "starting",
        Lifecycle::Running => "running",
        Lifecycle::Deactivated => "deactivated",
        Lifecycle::Quitting => "quitting",
        Lifecycle::Stopped => "stopped",
        Lifecycle::FailedOpen => "failed-open",
    };
    let view = match state.view() {
        View::Hidden => "hidden",
        View::Splash => "splash",
        View::Details => "details",
        View::Prompt { .. } => "prompt",
    };
    let mode = mode_name(state.mode());
    let root_stage = match state.root_stage() {
        RootStage::Initramfs => "initramfs",
        RootStage::Switching => "switching",
        RootStage::RealRoot => "real-root",
    };

    format!(
        "{{\"lifecycle\":\"{lifecycle}\",\"view\":\"{view}\",\"mode\":\"{mode}\",\"root_stage\":\"{root_stage}\",\"status\":{},\"message\":{},\"progress\":{}}}",
        json_optional_string(state.status()),
        json_optional_string(state.message()),
        state
            .progress()
            .map(|value| value.to_string())
            .unwrap_or_else(|| "null".to_owned()),
    )
}

pub fn mode_name(mode: Mode) -> &'static str {
    match mode {
        Mode::Boot => "boot",
        Mode::Shutdown => "shutdown",
        Mode::Reboot => "reboot",
        Mode::Update => "update",
        Mode::Upgrade => "upgrade",
    }
}

fn json_optional_string(value: Option<&str>) -> String {
    match value {
        Some(value) => format!("\"{}\"", json_escape(value)),
        None => "null".to_owned(),
    }
}

fn json_escape(value: &str) -> String {
    let mut escaped = String::with_capacity(value.len());
    for character in value.chars() {
        match character {
            '"' => escaped.push_str("\\\""),
            '\\' => escaped.push_str("\\\\"),
            _ => escaped.push(character),
        }
    }
    escaped
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::splash::root_transition::{RootTransitionError, SystemFailure};
    use crate::splash::state::PromptMetadata;
    use std::io;
    use std::path::{Path, PathBuf};

    fn running_state() -> SplashState {
        let mut state = SplashState::default();
        state.apply(StateAction::MarkRunning).unwrap();
        state
    }

    #[test]
    fn request_mapping_updates_only_presentation_state() {
        let mut state = running_state();
        handle_request(
            &mut state,
            &Frame::text(Opcode::Status, 1, "Mounting filesystems").unwrap(),
        )
        .unwrap();
        handle_request(
            &mut state,
            &Frame::text(Opcode::UpdateRootFs, 2, "/path/that/need/not/exist").unwrap(),
        )
        .unwrap();

        assert_eq!(state.status(), Some("Mounting filesystems"));
        assert_eq!(state.root_stage(), RootStage::RealRoot);
    }

    #[test]
    fn json_escapes_text_and_has_no_secret_field() {
        let mut state = running_state();
        state
            .apply(StateAction::SetMessage(Some(
                "quote: \" and slash: \\".into(),
            )))
            .unwrap();
        let json = state_json(&state);

        assert!(json.contains("quote: \\\" and slash: \\\\"));
        assert!(!json.contains("secret"));
        assert!(!json.contains("answer"));
    }

    #[test]
    fn only_observation_commands_are_non_mutating() {
        assert!(!is_mutating(Opcode::Ping));
        assert!(!is_mutating(Opcode::State));
        assert!(!is_mutating(Opcode::NativeReady));
        assert!(is_mutating(Opcode::Show));
        assert!(is_mutating(Opcode::Quit));
    }

    #[test]
    fn native_readiness_cannot_be_inferred_by_the_pure_command_handler() {
        let mut state = running_state();
        let response = handle_request(&mut state, &Frame::empty(Opcode::NativeReady, 27).unwrap())
            .unwrap()
            .response;

        assert_eq!(response.opcode(), Opcode::Error);
    }

    #[derive(Default)]
    struct RecordingTransition(Vec<PathBuf>);

    impl RootTransition for RecordingTransition {
        fn transition(&mut self, new_root: &Path) -> Result<(), RootTransitionError> {
            self.0.push(new_root.to_path_buf());
            Ok(())
        }
    }

    struct FailingTransition {
        error: RootTransitionError,
        calls: usize,
    }

    impl RootTransition for FailingTransition {
        fn transition(&mut self, _new_root: &Path) -> Result<(), RootTransitionError> {
            self.calls += 1;
            Err(self.error.clone())
        }
    }

    fn transition_failure(rollback_incomplete: bool) -> RootTransitionError {
        RootTransitionError::TransitionFailed {
            failure: SystemFailure {
                operation: "change directory to new root",
                kind: io::ErrorKind::Other,
                errno: None,
            },
            rollback_failures: rollback_incomplete
                .then_some(SystemFailure {
                    operation: "restore old root",
                    kind: io::ErrorKind::Other,
                    errno: None,
                })
                .into_iter()
                .collect(),
        }
    }

    #[test]
    fn root_handoff_runs_through_injected_adapter_seam() {
        let mut state = running_state();
        let mut transition = RecordingTransition::default();
        let request = Frame::text(Opcode::UpdateRootFs, 9, "/sysroot").unwrap();

        handle_request_with_root_transition(&mut state, &request, &mut transition).unwrap();
        handle_request_with_root_transition(&mut state, &request, &mut transition).unwrap();

        assert_eq!(transition.0, [PathBuf::from("/sysroot")]);
        assert_eq!(state.root_stage(), RootStage::RealRoot);
    }

    #[test]
    fn cleanly_rolled_back_root_failure_fails_open_and_stops_display_ownership() {
        let mut state = running_state();
        let mut transition = FailingTransition {
            error: transition_failure(false),
            calls: 0,
        };
        let request = Frame::text(Opcode::UpdateRootFs, 10, "/sysroot").unwrap();

        let outcome =
            handle_request_with_root_transition(&mut state, &request, &mut transition).unwrap();

        assert_eq!(outcome.response.opcode(), Opcode::Error);
        assert!(outcome.should_quit);
        assert!(
            outcome
                .fatal_root_transition
                .as_ref()
                .is_some_and(|error| !error.rollback_incomplete())
        );
        assert_eq!(state.lifecycle(), Lifecycle::FailedOpen);
        assert_eq!(state.root_stage(), RootStage::Switching);
        assert_eq!(transition.calls, 1);
    }

    #[test]
    fn incomplete_root_rollback_forces_failed_open_shutdown() {
        let mut state = running_state();
        let mut transition = FailingTransition {
            error: transition_failure(true),
            calls: 0,
        };
        let request = Frame::text(Opcode::UpdateRootFs, 11, "/sysroot").unwrap();

        let outcome =
            handle_request_with_root_transition(&mut state, &request, &mut transition).unwrap();

        assert_eq!(outcome.response.opcode(), Opcode::Error);
        assert!(outcome.should_quit);
        assert!(!outcome.retain_splash);
        assert!(
            outcome
                .fatal_root_transition
                .as_ref()
                .is_some_and(RootTransitionError::rollback_incomplete)
        );
        assert_eq!(state.lifecycle(), Lifecycle::FailedOpen);
        assert_eq!(state.root_stage(), RootStage::Switching);
        assert_eq!(transition.calls, 1);
    }

    #[test]
    fn root_handoff_rejection_during_prompt_fails_open_before_adapter_call() {
        let mut state = running_state();
        state
            .apply(StateAction::BeginPrompt(
                PromptMetadata::new(41, "Unlock volume").unwrap(),
            ))
            .unwrap();
        let mut transition = RecordingTransition::default();
        let request = Frame::text(Opcode::UpdateRootFs, 9, "/sysroot").unwrap();

        let outcome =
            handle_request_with_root_transition(&mut state, &request, &mut transition).unwrap();

        assert_eq!(outcome.response.opcode(), Opcode::Error);
        assert!(outcome.should_quit);
        assert!(!outcome.retain_splash);
        assert!(outcome.fatal_root_transition.is_none());
        assert!(transition.0.is_empty());
        assert_eq!(state.root_stage(), RootStage::Initramfs);
        assert_eq!(state.lifecycle(), Lifecycle::FailedOpen);
        assert!(state.view().prompt().is_none());
    }

    #[test]
    fn quit_carries_retain_intent_without_changing_protocol_response() {
        let mut state = running_state();
        let outcome = handle_request(&mut state, &Frame::quit(10, true).unwrap()).unwrap();

        assert!(outcome.should_quit);
        assert!(outcome.retain_splash);
        assert_eq!(outcome.response.opcode(), Opcode::Ack);
    }
}

File: src/splash/command.cpp
#include "sart/splash/command.hpp"

#include <algorithm>
#include <array>
#include <cstddef>
#include <fcntl.h>
#include <format>
#include <linux/openat2.h>
#include <optional>
#include <sys/stat.h>
#include <sys/syscall.h>
#include <system_error>
#include <unistd.h>

namespace sart::splash {
    namespace {

        std::string json_escape(std::string_view value) {
            std::string escaped;
            escaped.reserve(value.size());
            for (const auto character : value) {
                if (character == '"' || character == '\\')
                    escaped.push_back('\\');
                escaped.push_back(character);
            }
            return escaped;
        }

        std::string json_optional_string(const std::optional<std::string> &value) {
            return value ? std::format("\"{}\"", json_escape(*value)) : "null";
        }

        std::string_view lifecycle_name(Lifecycle lifecycle) {
            switch (lifecycle) {
            case Lifecycle::starting:
                return "starting";
            case Lifecycle::running:
                return "running";
            case Lifecycle::deactivated:
                return "deactivated";
            case Lifecycle::quitting:
                return "quitting";
            case Lifecycle::stopped:
                return "stopped";
            case Lifecycle::failed_open:
                return "failed-open";
            }
            return "failed-open";
        }

        std::string_view view_name(const View &view) {
            if (view.prompt_metadata())
                return "prompt";
            switch (*view.base_view()) {
            case BaseView::hidden:
                return "hidden";
            case BaseView::splash:
                return "splash";
            case BaseView::details:
                return "details";
            }
            return "hidden";
        }

        std::string_view root_stage_name(RootStage stage) {
            switch (stage) {
            case RootStage::initramfs:
                return "initramfs";
            case RootStage::switching:
                return "switching";
            case RootStage::real_root:
                return "real-root";
            }
            return "initramfs";
        }

        CommandOutcome state_error(std::uint64_t request_id, const std::exception &error) {
            return {Frame::error(request_id, error.what()), false, false, false};
        }

        class OwnedDescriptor {
          public:
            explicit OwnedDescriptor(int value = -1) : value_(value) {}
            OwnedDescriptor(const OwnedDescriptor &) = delete;
            OwnedDescriptor &operator=(const OwnedDescriptor &) = delete;
            ~OwnedDescriptor() {
                if (value_ >= 0)
                    close(value_);
            }
            [[nodiscard]] int get() const noexcept { return value_; }

          private:
            int value_;
        };

        [[noreturn]] void root_error(std::string_view operation) {
            throw std::system_error(errno, std::generic_category(), std::string(operation));
        }

        OwnedDescriptor secure_open(int directory, const char *path, std::uint64_t flags, std::uint64_t resolve) {
            open_how how{};
            how.flags = flags;
            how.resolve = resolve;
            const auto descriptor = static_cast<int>(syscall(SYS_openat2, directory, path, &how, sizeof(how)));
            if (descriptor < 0)
                root_error("securely open root transition object");
            return OwnedDescriptor(descriptor);
        }

        struct RootObject {
            dev_t device;
            ino_t inode;
            mode_t mode;
            uid_t uid;
            off_t size;
            auto operator<=>(const RootObject &) const = default;
        };

        RootObject inspect_root_object(int descriptor, bool directory, std::string_view name) {
            struct stat metadata{};
            if (fstat(descriptor, &metadata) != 0)
                root_error("inspect root transition object");
            if ((directory && !S_ISDIR(metadata.st_mode)) || (!directory && !S_ISREG(metadata.st_mode)) ||
                metadata.st_uid != 0 || (metadata.st_mode & 0022) != 0 || (metadata.st_mode & S_IXUSR) == 0 ||
                (!directory && (metadata.st_size <= 0 || metadata.st_size > 64 * 1024 * 1024))) {
                throw std::runtime_error("unsafe root transition object: " + std::string(name));
            }
            return {metadata.st_dev, metadata.st_ino, metadata.st_mode, metadata.st_uid, metadata.st_size};
        }

        void compare_executables(int candidate, int running, off_t size) {
            std::array<std::byte, 64 * 1024> left{};
            std::array<std::byte, 64 * 1024> right{};
            off_t offset{};
            while (offset < size) {
                const auto wanted = std::min<std::size_t>(left.size(), static_cast<std::size_t>(size - offset));
                const auto left_count = pread(candidate, left.data(), wanted, offset);
                const auto right_count = pread(running, right.data(), wanted, offset);
                if (left_count != static_cast<ssize_t>(wanted) || right_count != left_count) {
                    root_error("read root transition executable");
                }
                if (!std::equal(left.begin(), left.begin() + left_count, right.begin())) {
                    throw std::runtime_error("new-root sart differs from the running executable");
                }
                offset += left_count;
            }
        }

    } // namespace

    void DeferredRootTransition::transition(const std::filesystem::path &) {}

    void LinuxSelfRootTransition::transition(const std::filesystem::path &new_root) {
        const auto value = new_root.native();
        if (geteuid() != 0)
            throw std::runtime_error("root transition requires UID 0");
        if (!new_root.is_absolute() || value == "/" || value.empty() || value.size() >= 4096 || value.ends_with('/') ||
            value.find("//") != std::string::npos) {
            throw std::invalid_argument("candidate root path is not normalized and absolute");
        }
        for (const auto &component : new_root) {
            if (component == "." || component == "..") {
                throw std::invalid_argument("candidate root path contains a dot component");
            }
        }
        const auto root = secure_open(AT_FDCWD, value.c_str(), O_RDONLY | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW,
                                      RESOLVE_NO_SYMLINKS | RESOLVE_NO_MAGICLINKS);
        const auto root_before = inspect_root_object(root.get(), true, "candidate root");
        const auto usr = secure_open(root.get(), "usr", O_RDONLY | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW,
                                     RESOLVE_BENEATH | RESOLVE_NO_SYMLINKS | RESOLVE_NO_MAGICLINKS);
        const auto usr_before = inspect_root_object(usr.get(), true, "candidate usr");
        const auto bin = secure_open(usr.get(), "bin", O_RDONLY | O_DIRECTORY | O_CLOEXEC | O_NOFOLLOW,
                                     RESOLVE_BENEATH | RESOLVE_NO_SYMLINKS | RESOLVE_NO_MAGICLINKS);
        const auto bin_before = inspect_root_object(bin.get(), true, "candidate usr/bin");
        const auto candidate = secure_open(bin.get(), "sart", O_RDONLY | O_CLOEXEC | O_NOFOLLOW,
                                           RESOLVE_BENEATH | RESOLVE_NO_SYMLINKS | RESOLVE_NO_MAGICLINKS);
        const auto candidate_before = inspect_root_object(candidate.get(), false, "candidate sart");
        const OwnedDescriptor running(open("/proc/self/exe", O_RDONLY | O_CLOEXEC));
        if (running.get() < 0)
            root_error("open running executable");
        const auto running_before = inspect_root_object(running.get(), false, "running sart");
        if (candidate_before.size != running_before.size) {
            throw std::runtime_error("new-root sart size differs from the running executable");
        }
        compare_executables(candidate.get(), running.get(), candidate_before.size);
        if (inspect_root_object(root.get(), true, "candidate root") != root_before ||
            inspect_root_object(usr.get(), true, "candidate usr") != usr_before ||
            inspect_root_object(bin.get(), true, "candidate usr/bin") != bin_before ||
            inspect_root_object(candidate.get(), false, "candidate sart") != candidate_before ||
            inspect_root_object(running.get(), false, "running sart") != running_before) {
            throw std::runtime_error("root transition object changed during verification");
        }
        const OwnedDescriptor old_root(open("/", O_RDONLY | O_DIRECTORY | O_CLOEXEC));
        const OwnedDescriptor old_cwd(open(".", O_RDONLY | O_DIRECTORY | O_CLOEXEC));
        if (old_root.get() < 0 || old_cwd.get() < 0)
            root_error("save root transition rollback point");
        if (fchdir(root.get()) != 0)
            root_error("enter candidate root");
        if (chroot(".") != 0) {
            if (fchdir(old_cwd.get()) != 0) {
            }
            root_error("chroot to candidate root");
        }
        if (chdir("/") != 0) {
            if (fchdir(old_root.get()) != 0) {
            }
            if (chroot(".") != 0) {
            }
            if (fchdir(old_cwd.get()) != 0) {
            }
            root_error("change directory in candidate root");
        }
    }

    bool is_mutating(Opcode opcode) noexcept {
        return opcode != Opcode::ping && opcode != Opcode::state && opcode != Opcode::native_ready;
    }

    std::string_view mode_name(Mode mode) noexcept {
        switch (mode) {
        case Mode::boot:
            return "boot";
        case Mode::shutdown:
            return "shutdown";
        case Mode::reboot:
            return "reboot";
        case Mode::update:
            return "update";
        case Mode::upgrade:
            return "upgrade";
        }
        return "boot";
    }

    std::string state_json(const SplashState &state) {
        const auto progress = state.progress() ? std::to_string(*state.progress()) : std::string("null");
        return std::format("{{\"lifecycle\":\"{}\",\"view\":\"{}\",\"mode\":\"{}\",\"root_stage\":\"{}\",\"status\":{},"
                           "\"message\":{},\"progress\":{}}}",
                           lifecycle_name(state.lifecycle()), view_name(state.view()), mode_name(state.mode()),
                           root_stage_name(state.root_stage()), json_optional_string(state.status()),
                           json_optional_string(state.message()), progress);
    }

    CommandOutcome handle_request(SplashState &state, const Frame &request) {
        DeferredRootTransition transition;
        return handle_request(state, request, transition);
    }

    CommandOutcome handle_request(SplashState &state, const Frame &request, RootTransition &root_transition) {
        const auto request_id = request.request_id();
        if (request.opcode() == Opcode::ping) {
            return {Frame::pong(request_id), false, false, false};
        }
        if (request.opcode() == Opcode::state) {
            return {Frame::state_result(request_id, state_json(state)), false, false, false};
        }
        if (request.opcode() == Opcode::native_ready) {
            return {Frame::error(request_id, "native password broker is unavailable in this command context"), false,
                    false, false};
        }

        std::optional<StateAction> action;
        bool should_quit{};
        bool retain_splash{};
        switch (request.opcode()) {
        case Opcode::show:
            action = Show{};
            break;
        case Opcode::hide:
            action = Hide{};
            break;
        case Opcode::status: {
            const auto text = request.payload_text();
            action = SetStatus{text.empty() ? std::nullopt : std::optional{std::string(text)}};
            break;
        }
        case Opcode::progress:
            action = SetProgress{request.progress_value()};
            break;
        case Opcode::message:
            action = SetMessage{std::string(request.payload_text())};
            break;
        case Opcode::hide_message: {
            const auto requested = request.payload_text();
            if (requested.empty() || (state.message() && *state.message() == requested)) {
                action = SetMessage{std::nullopt};
            }
            break;
        }
        case Opcode::details_show:
            action = ShowDetails{};
            break;
        case Opcode::details_hide:
            action = HideDetails{};
            break;
        case Opcode::details_toggle:
            action = ToggleDetails{};
            break;
        case Opcode::deactivate:
            action = Deactivate{};
            break;
        case Opcode::reactivate:
            action = Reactivate{};
            break;
        case Opcode::set_mode:
            action = SetMode{*request.mode_value()};
            break;
        case Opcode::update_root_fs: {
            if (state.root_stage() == RootStage::initramfs) {
                try {
                    state.apply(SetRootStage{RootStage::switching});
                } catch (const StateError &error) {
                    try {
                        state.apply(FailOpen{});
                    } catch (...) {
                    }
                    auto outcome = state_error(request_id, error);
                    outcome.should_quit = true;
                    return outcome;
                }
            }
            if (state.root_stage() == RootStage::switching) {
                try {
                    root_transition.transition(std::filesystem::path(request.payload_text()));
                } catch (const std::exception &error) {
                    try {
                        state.apply(FailOpen{});
                    } catch (...) {
                    }
                    auto outcome = state_error(request_id, error);
                    outcome.should_quit = true;
                    outcome.fatal_root_transition = true;
                    return outcome;
                }
                action = SetRootStage{RootStage::real_root};
            }
            break;
        }
        case Opcode::quit:
            should_quit = true;
            retain_splash = request.retains_splash();
            action = Quit{};
            break;
        case Opcode::ack:
        case Opcode::error:
        case Opcode::pong:
        case Opcode::state_result:
            return {Frame::error(request_id, "response opcode is invalid in a request"), false, false, false};
        case Opcode::ping:
        case Opcode::state:
        case Opcode::native_ready:
            break;
        }
        if (action) {
            try {
                state.apply(std::move(*action));
            } catch (const StateError &error) {
                return state_error(request_id, error);
            }
        }
        return {Frame::ack(request_id), should_quit, retain_splash, false};
    }

} // namespace sart::splash
